_posts/2019-10-23-gpt-2/gpt2.Rmd

```diff
@@ -0,0 +1,317 @@
+
+Whenever a new token is to be predicted, a _softmax_ is taken over the vocabulary ^[For an introduction to how softmax activation behaves, see [Winner takes all: A look at activations and cost functions](https://blogs.rstudio.com/tensorflow/posts/2018-10-11-activations-intro/).]. Directly taking the softmax output amounts to maximum likelihood estimation. In reality, however, always choosing the maximum likelihood estimate results in highly repetitive output. 
+
+A natural option seems to be using the softmax outputs as probabilities: Instead of just taking the _argmax_, we sample from the output distribution. Unfortunately, this procedure has negative ramifications of its own. In a big vocabulary, very improbable words together make up a substantial part of the probability mass; at every step of generation, there is thus a non-negligible probability that an improbable word may be chosen. This word will now exert great influence on what is chosen next. In that manner, highly improbable sequences can build up. 
+
+The task thus is to navigate between the Scylla of determinism and the Charybdis of weirdness. With the GPT-2 model presented below, we have three options:
+
+- vary the _temperature_ (parameter `temperature`),
+- vary `top_k`, the number of tokens considered, or
+- vary `top_p`, the probability mass considered.
+
+The _temperature_ concept is rooted in statistical mechanics. Looking at the Boltzmann distribution, used to model state probabilities $p_i$dependent on energy $\epsilon_i$:
+
+$$p_i \sim e^{-\frac{\epsilon_i}{kT}}$$
+
+we see there is a moderating variable _temperature_ $T$ ^[$k$ is the Boltzmann constant] that dependent on whether it's below or above 1, will exert an either amplifying or attenuating influence on differences between probabilities.
+
+Analogously, in the context of predicting the next token, the individual logits are scaled by the temperature, and only then is the softmax taken. Temperatures below zero would make the model even more rigorous in choosing the maximum likelihood candidate; instead, we'd be interested in experimenting with temperatures above 1 to give higher chances to less likely candidates -- hopefully, resulting in more human-like text.
+
+In top-$k$ sampling, the softmax outputs are sorted, and only the top-$k$ tokens are considered for sampling. The difficulty here is how to choose $k$. Sometimes a few words make up for almost all probability mass, in which case we'd like to choose a low number; in other cases the distribution is flat, and a higher number would be adequate.
+
+This sounds like rather than the number of candidates, a target probability mass should be specified. This is the approach suggested by [@2019arXiv190409751H]. Their method, called top-$p$, or Nucleus, sampling, computes the cumulative distribution of softmax outputs and picks a cut-off point $p$. Only the tokens constituting the top-$p$ portion of probability mass is retained for sampling.
```

Comment:
```suggestion
This sounds like rather than the number of candidates, a target probability mass should be specified. This is the approach suggested by [@2019arXiv190409751H]. Their method, called top-$p$, or Nucleus sampling, computes the cumulative distribution of softmax outputs and picks a cut-off point $p$. Only the tokens constituting the top-$p$ portion of probability mass is retained for sampling.
```